### PR TITLE
Dist hess schur

### PR DESCRIPTION
**Pull request by poulson**
Adding preliminary support for distributed (real and complex; multi-bulge and AED) Hessenberg QR algorithms. These algorithms have undergone modest testing and have not yet been performance tuned.

### CODE REVIEW

File: include/El/blas_like/level1/Copy/GeneralPurpose.hpp [hunk 1]
```diff
@@ -93,19 +98,24 @@ void Helper
     if( includeViewers )
     {
         comm = g.ViewingComm();
-        const int commSize = mpi::Size( comm );
+        const int viewingSize = mpi::Size( g.ViewingComm() );
+        const int distBSize = mpi::Size( B.DistComm() );
 
-        vector<int> distMap(commSize);
-        for( int q=0; q<commSize; ++q )
+        vector<int> distBToViewing(distBSize);
```
Comment (rhl-) [inline include/El/blas_like/level1/Copy/GeneralPurpose.hpp]: std::vector<int>

File: include/El/blas_like/level1/Copy/GeneralPurpose.hpp [hunk 2]
```diff
@@ -93,19 +98,24 @@ void Helper
...
+        for( int distBRank=0; distBRank<distBSize; ++distBRank )
```
Comment (rhl-) [inline include/El/blas_like/level1/Copy/GeneralPurpose.hpp]: why `int` and not `Int`

File: src/core/DistMatrix/Abstract.cpp [hunk 3]
```diff
@@ -582,42 +582,43 @@ void AbstractDistMatrix<T>::ProcessQueues( bool includeViewers )
...
+    // We will first push to redundant rank 0
+    const int redundantRoot = 0;
+
     // Compute the metadata
     // ====================
     mpi::Comm comm;
     vector<int> sendCounts, owners(totalSend);
     if( includeViewers )
     {
         comm = g.ViewingComm();
```
Comment (rhl-) [inline src/core/DistMatrix/Abstract.cpp]: same.

File: include/El/blas_like/level1/Copy/GeneralPurpose.hpp (hunk 1 above)
Comment (poulson) [inline include/El/blas_like/level1/Copy/GeneralPurpose.hpp]: Actually, Elemental imports `std::vector` as `El::vector` here: https://github.com/elemental/Elemental/blob/1bd259d51aa1e7f7ed40306580a302640c36b2d6/include/El/core/environment/decl.hpp#L19

File: include/El/blas_like/level1/Copy/GeneralPurpose.hpp (hunk 2 above)
Comment (poulson) [inline include/El/blas_like/level1/Copy/GeneralPurpose.hpp]: MPI forces ranks to be indexed with `int`, so this is actually correct

**Comment by poulson** [comment]
I am going forward and merging since this fixes several outstanding bugs in `DistMatrix<T,MC,MR,BLOCK>`

**Pull request merged**